ENT-16 Make SSO configurable per Site

bdero opened the pull request:
This change adds a Site field to the ProviderConfig models and enforces each SSO integration to only work for a specific Site (or no Site if unset).

**JIRA tickets**: Implements [ENT-16](https://openedx.atlassian.net/browse/ENT-16).

**Discussions**: This was initially discussed in the comments of [ENT-13](https://openedx.atlassian.net/browse/ENT-13) and formed into it's own ticket: [ENT-16](https://openedx.atlassian.net/browse/ENT-16).

**Sandbox URL**:

SITE 1: http://pr13474.sandbox.opencraft.hosting/
SITE 2: http://pr13474.sandbox.bdero.me/

**Testing instructions**:

1. Navigate to both [this sandbox login](http://pr13474.sandbox.opencraft.hosting/login) and [this second sandbox login](http://pr13474.sandbox.bdero.me/login) in separate tabs. Notice that they're the same sandbox, but the two URLs are configured to forge different Sites.

   Observe that the first Site only shows the normal login screen, while the second one shows a "TestShib College" login.
1. Login using "TestShib College". Observe that it still works.
1. As a superuser, navigate to http://pr13474.sandbox.bdero.me/admin/third_party_auth/samlconfiguration/.
1. Note that the SAML Configuration admin view shows the `Site` and `Update` columns, as well as the filter section (which shows up for keyed models).

   ![](http://i.imgur.com/t613Bx5.png)
1. Navigate to http://pr13474.sandbox.opencraft.hosting/auth/saml/metadata.xml and http://pr13474.sandbox.bdero.me/auth/saml/metadata.xml.

   - Observe that the entityID within the EntityDescriptor corresponds with the entityID of the `SAMLConfiguration` models for the different Sites.
   - Observe that the certificates are different.
   - Observe that in both cases, the domain of the URI in the `Location` attribute of `AssertionCustomerService` matches the domain of the current URL.

**Merge deadline**: **9/22/2016**

**Author notes and concerns**:

Making it configurable by site and having only the current Site's login options show up was trivial, but I couldn't find an sane universal way to disable `/auth/login/<backend>` redirects, but users trying to login using the wrong Site's provider shouldn't be able to make it all the way through the login process (and OAuth2/LTI require a matching callback to the referrer).

**Reviewers**
- [x] @e-kolpakov 
- [ ] edX reviewer: @douglashall 

**Settings**
```yaml
EDXAPP_FEATURES:
  ENABLE_COMBINED_LOGIN_REGISTRATION: true
  ENABLE_THIRD_PARTY_AUTH: true
```

douglashall commented:
@bdero Yes, this looks like a good approach. Happy to give a full review when you are ready. Thanks.

douglashall commented:
FYI @mattdrayer

mattdrayer commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -226,7 +237,27 @@ def display_for_login(self):
+        return self.site is None
+
+    @property
+    def usable(self):
```
Since both of these properties are new, does it make sense to combine `usable` with `for_current_site` to yield `enabled_for_current_site`?

bdero replied:
Yes, I've combined them. I initially wrote them separately as I anticipated needing to use both of them, but it turns out that `usable` is the only check necessary

mattdrayer commented on common/djangoapps/third_party_auth/saml.py:
```diff
@@ -41,10 +41,18 @@ def auth_url(self):
             raise Http404
+
         # TODO: remove this check once the fix is merged upstream:
         # https://github.com/omab/python-social-auth/pull/821
```
Unrelated to this PR, but it appears https://github.com/omab/python-social-auth/pull/821 has been merged.

bdero replied:
@mattdrayer Nice, thanks for checking. Do you think we should remove that hack and bump the python-social-auth version in a later PR?

mattdrayer replied:
It would be an act of good citizenry, sure.  Happy to review the PR if you post one.

bdero commented:
jenkins test js

bdero commented:
@e-kolpakov This is all set for review now.

bradenmacdonald commented:
@bdero If one goes to /auth/saml/metadata.xml on each different site, does it return different XML? That could be a missing piece here. I believe the SAML metadata XML contains a URL for the site, and it definitely contains an Entity ID which is usually a URL. So I think you may need to make `SAMLConfiguration` site-keyed as well.

e-kolpakov commented on common/djangoapps/third_party_auth/saml.py:
```diff
@@ -41,10 +41,18 @@ def auth_url(self):
         if 'idp' not in self.strategy.request_data():
             raise AuthMissingParameter(self, 'idp')
+
+        from .models import SAMLProviderConfig
```
@bdero just to make sure - is local import required here?

bdero replied:
Yes, I didn't look too hard into it, but the python path of this module at the time of execution doesn't contain this directory nor the common djangoapps when this module is loaded (I think this has to do with when python-social-auth djangoapp is initialized).
There's another preexisting location in this file which performs this import the same way for (presumably) the same reason.

e-kolpakov replied:
@bdero I might be misusing the term here, but I was talking about putting this import at the top of the file, rather than inside a method. If I'm not mistaking importing something inside method is called local import, while it seems to me you have tried to advocate for relative import (the one that starts with `.` rather than top-level package).

bdero replied:
Yup, that's what I meant. Neither the relative nor top-level imports worked when the module was first loaded, so I stuck with doing a local import of `SAMLProviderConfig`, just like [`get_idp()`](https://github.com/edx/edx-platform/pull/13474/files#diff-31c7777edd76b5f6c6190122337c991cL22) does.

e-kolpakov commented on common/djangoapps/third_party_auth/tests/test_provider.py:
```diff
@@ -84,6 +90,41 @@ def test_providers_displayed_for_login(self):
+        default_site_provider = self.configure_google_provider(visible=True, enabled=True)
+        default_site_disabled_provider = self.configure_google_provider(visible=True, enabled=False)
+        site_provider = self.configure_google_provider(visible=True, enabled=True, site=site_a)
+        site_disabled_provider = self.configure_google_provider(visible=True, enabled=False, site=site_a)
```
@bdero `configure_google_provider` creates an `OAuth2ProviderConfig` and if I recall right `OAuth2ProviderConfig` is a versioned model (inherited from `ConfigurationModel`), so any updates/insertes to the entity with same "key" (site and provider name in this case) from app point of view update the "current" record. So, you might not need to do it twice: `self.configure_google_provider(visible=True, enabled=False, site=site_a)` should be enough. 

But I might be wrong.

bdero replied:
@e-kolpakov Yes, I think you're correct here (even though it seemed to be working in this context before). All of the ProviderConfig models are keyed to the Site now, and this test is now reconfiguring the provider between each DDT test case, so I think each case should be sufficiently isolated now.

e-kolpakov commented on common/djangoapps/third_party_auth/tests/test_provider.py:
```diff
@@ -84,6 +90,41 @@ def test_providers_displayed_for_login(self):
+            for auth_provider, value in assertions:
+                self.assertEqual(auth_provider.enabled_for_current_site, value)
+
+        enabled_for_current_site((
```
@bdero looks like you're trying to evaluate a number of cases in single test. Usually we (both edX and Opencraft) tend to write such kind of tests as parameterized tests using `ddt`. The benefit of doing so is that while still sharing the code, there are one test for every case, so when some of the conditions are not met you get a better understanding what exactly went wrong (which pair caused an error) and all the other cases are tested too.

In this case, you should have something like this:

    @ddt.data(
        (True, None, True),
        (False, None, False),
        (True, site_a, True),
        (False, site_a, False),
        ....
    )
    @ddt.unpack
    test_providers_enabled_for_site(enabled, site, expected_enabled):
        provider = self.configure_google_provider(visible=True, enabled=enabled, site=site)
        self.assertEqual(provider.enabled_for_current_site, expected_enabled)

   # and the same one wrapped into with_site_configuration_context(SITE_DOMAIN_A)

bdero replied:
Good idea, I refactored this test to use DDT.

e-kolpakov commented:
@bdero For some reason third party auth does not work well in my devstack - despite setting FEATURE flag, configuring Testshib IdP and pulling metadata it still shows no third party auth options (not even google/oauth). I don't have time to debug this issue right now, so I would appreciate if you could configure sandbox for testing - I'll verify it on sandbox than.

bdero commented:
@bradenmacdonald @e-kolpakov Keying the models by Site turned out to be hairy. Django started producing ambiguous queries when viewing the provider admin views and I tracked it down to the queryset extensions [I fixed in this commit](https://github.com/edx/edx-platform/pull/13474/commits/7a96f2b1267ff9ab31cf25236ebebf52b7bd512b).

Note the change (for future reviews):
> This change disambiguates the use of `id` in the
> ConfigurationModelManager queryset extensions. This is required
> when specifying KEY_FIELDS where there is more than one key and
> at least one of them is a ForeignKey. The table must be specified
> in this case because the query extension may be appended to the
> WHERE clause of queries selecting `id` fields from multiple tables
> via an INNER JOIN clause.

I think I've gotten by most of the roadblocks at this point, but probably this will need more work next week and I'll be picking this up again on Monday.

bradenmacdonald commented:
@bdero Interesting - thanks for figuring that out and fixing it.

e-kolpakov commented on common/djangoapps/third_party_auth/provider.py:
```diff
@@ -16,18 +17,19 @@ class Registry(object):
+        site = get_current_site()
         for backend_name in _PSA_OAUTH2_BACKENDS:
-            provider = OAuth2ProviderConfig.current(backend_name)
+            provider = OAuth2ProviderConfig.current(get_current_site(), backend_name)
```
@bdero nit: `get_current_site` is already saved in `site` variable two lines above.

e-kolpakov commented on common/djangoapps/third_party_auth/tests/test_provider.py:
```diff
@@ -90,40 +93,23 @@ def test_providers_displayed_for_login(self):
-                (site_disabled_provider, False),
-                (site_b_provider, False),
-            ))
+            self.assertEqual(provider.enabled_for_current_site, expected_a)
```
@bdero nit: might make sense to extract this into a dedicated test

bdero replied:
I've refactored these tests now so that they're two distinct tests.

e-kolpakov commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -161,7 +161,11 @@ def clean(self):
         """ Unique string key identifying this provider. Must be URL and css class friendly. """
         assert self.prefix is not None
-        return "-".join((self.prefix, ) + tuple(getattr(self, field) for field in self.KEY_FIELDS))
+        fields = []
```
@bdero nit: might make sense to initialize `fields` list with `self.prefix`, i.e.

    fields = [self.prefix]
    ....
    return "-".join(fields)

On the other hand you might want to keep the semantics of `fields` list keeping actual field values. So totally up to you :)

bdero replied:
Thankfully, this change is no longer necessary as I took a different approach. Modifying provider_id was a bit too risky anyway.

e-kolpakov commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -473,8 +467,6 @@ class SAMLConfiguration(ConfigurationModel):
     )
     site = models.ForeignKey(
         Site,
-        null=True,
```
@bdero If I recall right `null` in site column was used to mark providers enabled for all sites. How would it work with this version? I mean, is "global" provider configs no longer supported?

bdero replied:
This is no longer here. None of the site keys are nullable, but they default to the site specified in settings.SITE_ID.

e-kolpakov commented on common/djangoapps/config_models/models.py:
```diff
@@ -41,7 +41,10 @@ def current_set(self):
         assert self.model.KEY_FIELDS != (), "Just use model.current() if there are no KEY_FIELDS"
         return self.get_queryset().extra(           # pylint: disable=no-member
-            where=["id IN ({subquery})".format(subquery=self._current_ids_subquery())],
+            where=["{table_name}.id IN ({subquery})".format(
```
@bdero this `{table_name}.id IN ({subquery}).format()` occurs three times in this file - might make sense to extract it into a helper method (if possible)

bdero replied:
It actually only occurs twice, and the `subquery` is different between the two. I feel it may be overkill in this case to factor this out into a separate method as such a change would result in more code, not less.

Do you think it makes sense to leave it as-is?

e-kolpakov replied:
@bdero sure, that's just a nitpick (though for some reason I haven't mentioned that)

e-kolpakov commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -243,15 +244,7 @@ def enabled_for_current_site(self):
-            # If there's a site enabled for this request, then return True if the site matches.
-            config = configuration_helpers.get_current_site_configuration()
-            return config.site == self.site
-        # If not viewing a site, return True if the site is unset (which signifies that the provider is intended for
```
@bdero same question here as non nullable `site` column - is this feature (global provider for all sites) removed (or never existed?)

bdero replied:
This code is no longer here. It's much simpler now that the sites aren't nullable. :)

e-kolpakov commented on common/djangoapps/config_models/models.py:
```diff
@@ -136,9 +145,9 @@ def current(cls, *args):
 
     @classmethod
-    def is_enabled(cls):
+    def is_enabled(cls, *args):
```
@bdero `is_enabled` is almost always called with single parameter `site` - do you think it would make sense to promote it to an actual parameter? Also, it might be good to give it a default value of current site. SOmething like

    def is_enabled(site=None, * args):
       site = site if site else get_current_site()
       return cls.current(site, * args),enabled

bdero replied:
I don't think so, since `ConfigurationModel` is the abstract base for many models that have no context for Sites, such as `DashboardConfiguration`.

e-kolpakov replied:
@bdero Ok.

Another alternative would be to create another subclass, like `SiteDependentConfigurationModel`, but that might be an overkill for this use case. So I'm fine with what we have now.

bdero replied:
Yeah, I think this would have made sense with the last approach given I was piping it through all over the third_party_auth code base. Luckily it only happens in two or three places now including tests. :)

e-kolpakov commented on common/djangoapps/third_party_auth/pipeline.py:
```diff
@@ -82,6 +82,7 @@ def B(*args, **kwargs):
 
 from logging import getLogger
 
+
```
@bdero extra newline in otherwise unmodified file - please remove.

bdero replied:
Oops, I wonder how that got there. I removed it. Thanks. :)

e-kolpakov commented:
@bdero reviewed the code, left some comments and suggestions. Also:

1. Sandbox throws 500 error - please fix.
2. Please fix CI build

bdero commented:
I've been having a hard time getting the test environment to work, but I've gotten around a third of the tests to pass by mocking site lookups appropriately.

I'm currently stuck on a problem where many tests are attempting to create multiple users with the same username. This might indicate that there's something wrong with my implementation, but I'm having trouble understanding what the intended behavior is.

```pytb
Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/case.py", line 331, in run
    testMethod()
  File "/edx/app/edxapp/edx-platform/common/djangoapps/third_party_auth/tests/specs/base.py", line 794, in test_first_party_auth_trumps_third_party_auth_and_fails_when_credentials_bad
    email='user@example.com', password='password', success=False)
  File "/edx/app/edxapp/edx-platform/common/djangoapps/third_party_auth/tests/specs/base.py", line 340, in assert_first_party_auth_trumps_third_party_auth
    strategy, email, password, self.get_username(), skip_social_auth=True)
  File "/edx/app/edxapp/edx-platform/common/djangoapps/third_party_auth/tests/specs/base.py", line 461, in create_user_models_for_existing_account
    user = social_utils.Storage.user.create_user(email=email, password=password, username=username)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/social/storage/django_orm.py", line 60, in create_user
    return cls.user_model().objects.create_user(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/contrib/auth/models.py", line 187, in create_user
    **extra_fields)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/contrib/auth/models.py", line 182, in _create_user
    user.save(using=self._db)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/base.py", line 734, in save
    force_update=force_update, update_fields=update_fields)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/base.py", line 762, in save_base
    updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/base.py", line 846, in _save_table
    result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/base.py", line 885, in _do_insert
    using=using, raw=raw)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/manager.py", line 127, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/query.py", line 920, in _insert
    return query.get_compiler(using=using).execute_sql(return_id)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 974, in execute_sql
    cursor.execute(sql, params)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/utils.py", line 98, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/backends/sqlite3/base.py", line 318, in execute
    return Database.Cursor.execute(self, query, params)
IntegrityError: column username is not unique
```

For example, the problem manifests [in this test](https://github.com/edx/edx-platform/blob/408d7a86751d6920fd3f5607f8479a6b7b345a23/common/djangoapps/third_party_auth/tests/specs/base.py#L681-L712) among others:
- The first time a User is created by the social auth pipeline when [self.fake_auth_complete(strategy)](https://github.com/edx/edx-platform/blob/408d7a86751d6920fd3f5607f8479a6b7b345a23/common/djangoapps/third_party_auth/tests/specs/base.py#L681-L712) is called. **I think this may be the issue (?)** since I suspect it's supposed to redirect the new user to the registration form rather than actually creating the user model right away.
- The second time a user is created is [on the line below](https://github.com/edx/edx-platform/blob/408d7a86751d6920fd3f5607f8479a6b7b345a23/common/djangoapps/third_party_auth/tests/specs/base.py#L681-L712) and is of course also done by the social auth pipeline:

   ```python
user = self.create_user_models_for_existing_account(
        strategy, 'user@example.com', 'password', self.get_username())
```

Does anyone with context for this part of the codebase have any thoughts about what might be going wrong here?

bradenmacdonald commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -288,7 +306,7 @@ class SAMLProviderConfig(ProviderConfig):
     """
     prefix = 'saml'
-    KEY_FIELDS = ('idp_slug', )
+    KEY_FIELDS = ('site_id', 'idp_slug',)
```
For `SAMLProviderConfig`, I don't think we actually need to make `site_id` a `KEY_FIELD`. Instead, administrators can simply set a different `idp_slug` for each site (`university1-default`, `university2-othersite`).

Same might apply for LTI.

For OAuth2, it's necessary because OAuth2 providers are currently keyed by backend name (e.g. Google) rather than some sort of custom identifier (so if we didn't key by site, there would be no other way to have, say, Google on multiple sites). However, you could change that by adding a new `provider_slug` field to OAuth2Provider (with a data migration that sets it to the `backend_name` for backwards compatibility). Then, the unique provider_id for OAuth2Providers could be "oa2-[provider_slug]" instead of "oa2-[backend_name]". This would be more flexible and more consistent with SAML.

With such a change, you wouldn't need to make `site_id` a KEY_FIELD for any of the providers, and this PR would get a lot simpler. You wouldn't need to change the shared ConfigurationModel code at all.

bdero replied:
Yes, I like this idea much better than adding the site keys. I reverted most of the changes I've made since doing so and will try to take this approach instead.

bradenmacdonald commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -149,7 +160,7 @@ def clean(self):
         """ Unique string key identifying this provider. Must be URL and css class friendly. """
         assert self.prefix is not None
-        return "-".join((self.prefix, ) + tuple(getattr(self, field) for field in self.KEY_FIELDS))
+        return "-".join((self.prefix, ) + tuple(str(getattr(self, field)) for field in self.KEY_FIELDS))
```
Note that provider_ids are sometimes used in URLs, so changing the provider_id could cause breaking changes.

For example, university partners can currently do a "hinted login" by sending their users to a URL like [https://edge.edx.org/courses/course-v1:UBC+CourseName+2016/courseware/**?tpa_hint=saml-ubc-edge**](https://edge.edx.org/courses/course-v1:UBC+CourseName+2016/courseware/?tpa_hint=saml-ubc-edge) (Open that example in an incognito window to see), where the argument is the `provider_id`. So if you change the format of the `provider_id`, such links would stop working.

My suggestion (see previous comment) is that we don't make `site_id` a key field, and that we give OAuth2 providers a new "slug" field that defaults to the backend_name, which would preserve backward compatibility.

bradenmacdonald commented:
@bdero The tests in `IntegrationTest` are a bit fragile and confusing. That's why I made the alternative `IntegrationTestMixin` class which IMHO is easier to understand, and use that for newer integration tests.

To help debug that issue, I would suggest you checkout the commit from prior to your changes, and check whether or not `self.fake_auth_complete(strategy)` is creating a user at that point; then you can better understand how the test is supposed to work.

bdero commented:
jenkins test bokchoy

bdero commented:
jenkins test lettuce

e-kolpakov commented:
@bdero just a reminder - there are some unaddressed comments from my previous review - you might want to do something about them (i.e. change code or at say that you won't do that and why)

bdero commented:
@e-kolpakov This PR has gone through a major refactor as there were issues with the approach that @bradenmacdonald pointed out. Most of your comments address implementation details and nits. I will be addressing your comments once it's possible for me to know which ones will still be relevant once the feature works.

bdero commented:
jenkins test python

bdero commented:
I'm having trouble uploading seemingly valid metadata to TestShib, so I sent an email to the Shibboleth Users list about it: http://shibboleth.1660669.n2.nabble.com/Issues-uploading-metadata-XML-to-TestShib-td7628451.html

bdero commented:
@e-kolpakov I've addressed your comments, fixed the majority of the test failures, updated the sandbox, and wrote new testing instructions for the sandbox. Can you give this another review? Unfortunately, the issue with TestShib still isn't resolved, so I can't upload metadata and it's therefore not possible to complete a login. However, the bulk of the changes can be tested without actually completing a Shibboleth login.

CC @douglashall It might be good to start reviewing this, since it's in the current sprint and the merge deadline is this week.

bdero commented:
Update: Uploading metadata on TestShib is working again! Now it's possible to complete the testing on the sandbox.

e-kolpakov commented:
@bdero :+1: conditional CI build is fixed and docs team is pinged to update TPA docs.

- [x] I tested this: different sites backed by the same instance show different lists of TPA providers and different SAML configs.
- [x] I read through the code
- [ ] Includes documentation

bdero commented:
@douglashall This is ready for your review. :)

The sandbox is working and all tests are passing (unless the rebase I just performed screws something up). I've also squashed + did some reordering/edits and now this PR consists of two separate commits:

1. **Disambiguate queryset extensions for ConfigurationModels:** This is a change I had to make to the ConfigurationModel queryset extensions in order to support Foreign keys as KEY_FIELDS. It has a detailed commit description explaining why it's necessary.
1. **Implement Site settings for Third Party Auth:** This is the feature implementation mapping all Third Party Auth providers to Sites.

bdero commented:
CC @catong: Are you the right person for me to ping about needed updates to the TPA docs? :)

catong commented:
@bdero Thank you for the tag! We haven't assigned a writer to this project yet. What stage in the project are you at, so that I can understand how we can help at this point?

douglashall commented on common/djangoapps/third_party_auth/provider.py:
```diff
@@ -18,17 +19,17 @@ def _enabled_providers(cls):
         """ Helper method to iterate over all providers """
```
Update docstring to indicate that this returns an iterator over all providers for the current site.

bdero replied:
Done.

douglashall commented on common/djangoapps/third_party_auth/provider.py:
```diff
@@ -84,16 +85,16 @@ def get_enabled_by_backend_name(cls, backend_name):
         if backend_name in _PSA_OAUTH2_BACKENDS:
             provider = OAuth2ProviderConfig.current(backend_name)
-            if provider.enabled:
+            if provider.enabled_for_current_site:
```
Update the docstring of this function to indicate you are returning a generator for over providers for the current site.

bdero replied:
Done.

douglashall commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -226,7 +236,14 @@ def display_for_login(self):
+        """
+        Determines if the provider is able to be used with the current site.
+        """
+        return bool(self.enabled and self.site == get_current_site_default())
```
No need to wrap this in `bool`.

bdero replied:
Thanks, I removed it.

douglashall commented on common/djangoapps/third_party_auth/views.py:
```diff
@@ -36,7 +37,7 @@ def saml_metadata_view(request):
     You must send this XML to any Shibboleth Identity Provider that you wish to use.
     """
-    if not SAMLConfiguration.is_enabled():
+    if not SAMLConfiguration.is_enabled(get_current_site_default()):
```
No need to use `get_current_site_default` here. You can use the request object to get the current site, `request.site`.

bdero replied:
Done.

douglashall commented on openedx/core/djangoapps/theming/helpers.py:
```diff
@@ -154,6 +155,19 @@ def get_current_site():
+    site = get_current_site()
+    if site is None:
+        return Site.objects.get(id=settings.SITE_ID)
+    return site
```
You don't need this function. We are currently [monkey-patching](https://github.com/edx/edx-django-sites-extensions/blob/master/django_sites_extensions/models.py#L25) the Django version of `django.contrib.sites.models.SiteManager.get_current` to provide this desired functionality. You can remove this and replace all references with `Site.objects.get_current()`.

douglashall commented on common/djangoapps/third_party_auth/saml.py:
```diff
@@ -41,10 +42,18 @@ def auth_url(self):
+        provider = SAMLProviderConfig.current(self.strategy.request_data()['idp'])
+        if not provider.enabled_for_current_site:
+            log.error('The SAML authentication provider is disabled for this site')
+            raise Http404
```
Looks like this if clause is not [covered](https://build.testeng.edx.org/job/edx-platform-python-unittests-pr/25582/Diff_Coverage_Report/). Any chance you can add coverage?

bdero replied:
Thanks, I added a test that should cover this code. The test suite can take quite a while to startup after first booting the VM (even with `--disable-migrations` and `--fast`, so my tests are still booting locally for me. Despite this, I pushed the changes and hopefully this test will pass/the coverage will increase to 100% for that file.

douglashall commented:
@bdero Can you set `ENABLE_COMBINED_LOGIN_REGISTRATION` for the http://pr13474.sandbox.bdero.me/ site so that we can see the separate list of idp's per site?

bdero commented on common/djangoapps/third_party_auth/tests/test_provider.py:
```diff
@@ -1,10 +1,16 @@
 import unittest
 
+SITE_DOMAIN_A = 'mitprofessionalx.mit.edu'
+SITE_DOMAIN_B = 'somethingelse.mit.edu'
```
Done.

bdero commented on common/djangoapps/third_party_auth/tasks.py:
```diff
@@ -36,7 +37,7 @@ def fetch_saml_metadata():
         num_total: Total number of providers whose metadata was fetched
     """
-    if not SAMLConfiguration.is_enabled():
+    if not SAMLConfiguration.is_enabled(get_current_site_default()):
```
This task runs during times when there is no request context (saml management command, and I also think it runs on startup), so the default site fallback is necessary here.

douglashall replied:
Okay, seems like we would never be able to run this then for sites other than the default site. Does this matter? It seems IdPs associated with site other than the default site would never have their metadata updated. Let me know if I am missing something.

bdero replied:
Oops! Yes, you are right, what we really need to do is run this for any SAMLProviderConfig with a matching SAMLConfiguration enabled for it's Site.

I've made the fix and pushed.

By the way, I took a look and I don't think this is run at startup, it's run whenever a SAMLConfiguration is updates, and also the "saml pull" management command.

bdero commented on common/djangoapps/config_models/models.py:
```diff
@@ -136,9 +145,9 @@ def current(cls, *args):
     @classmethod
-    def is_enabled(cls):
+    def is_enabled(cls, *args):
         """Returns True if this feature is configured as enabled, else False."""
```
Done.

bdero commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -226,7 +236,14 @@ def display_for_login(self):
+        return bool(self.enabled_for_current_site and self.accepts_logins and self.visible)
+
+    @property
+    def enabled_for_current_site(self):
```
I'm willing to abstract an `enabled_for_site` out of this, which `enabled_for_current_site` uses, but there are many places that specifically need to check if a ProviderConfig is enabled against the current site.

Should I keep `enabled_for_current_site` and have it call the `enabled_for_site` property (using `Site.objects.get_current()` to fetch the current Site, as you mentioned in comments below)?

douglashall replied:
Eh, nevermind. I would keep this as `is_enabled_for_current_site()` and have it use `Site.objects.get_current()`. Thanks.

bdero commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -244,6 +261,12 @@ class OAuth2ProviderConfig(ProviderConfig):
             # To be precise, it's set by AUTHENTICATION_BACKENDS - which aws.py sets from THIRD_PARTY_AUTH_BACKENDS
         )
     )
+    provider_slug = models.SlugField(
```
The `LTIProviderConfig` identifies itself using the `lti_consumer_key`, which is a 255 character string. I think it makes sense to key LTI this way and not with a separate slug, so I feel that abstracting this into the common `ProviderConfig` might not be the right solution. Thoughts?

douglashall replied:
Got it. I missed the fact that `LTIProviderConfig` also extends `ProviderConfig`. Thanks for the clarification. Let's stick with what you have.

bdero commented:
@douglashall Do you mean change the site configuration to look like this?
![](http://i.imgur.com/1tjDA8n.png)

I was under the impression that it would be enabled for all sites by default when using the LMS config, but there might be something for me to learn here. :)

bdero commented:
@douglashall I have addressed all of your comments - contingent on the test suite passing this should be ready for another pass!

douglashall commented on openedx/core/djangoapps/theming/helpers.py:
```diff
@@ -6,6 +6,7 @@
 from path import Path
 
 from django.conf import settings, ImproperlyConfigured
+from django.contrib.sites.models import Site
```
Unused import.

bdero replied:
Thanks, I fixed this and pushed.

douglashall commented:
@bdero Thanks!

douglashall commented on common/djangoapps/third_party_auth/tasks.py:
```diff
@@ -36,16 +37,13 @@ def fetch_saml_metadata():
     for idp_slug in SAMLProviderConfig.key_values('idp_slug', flat=True):
         config = SAMLProviderConfig.current(idp_slug)
-        if not config.enabled:
+        if not config.enabled or SAMLConfiguration.is_enabled(config.site):
```
Should this be `if not config.enabled or not SAMLConfiguration.is_enabled(config.site):`?

bdero replied:
Yup. :) I believe this caused the test failures as well. I fixed and pushed.

bdero commented:
jenkins test python

bdero commented:
@douglashall Can you give this one final look and merge this? (I don't have write access to edx-platform)

I removed the uncovered lines as the check I added was actually redundant (because SAMLConfigurations are keyed to sites).

bdero commented:
@catong Thanks! I think the status of this feature is that it's ready to merge. Review has covered the high level approach as well as implementation details to the extent where there shouldn't be any significant changes from here.

douglashall commented:
@mattdrayer You okay with me merging this or did you want to take a review pass and merge yourself?

mattdrayer commented:
:shipit:

bdero commented:
:tada:
